### PR TITLE
Initial FrSky ACCST V2.1.x support

### PR DESCRIPTION
**Pull request by MJ666**
Tested on bench with a Devo10 and an R-XSR receiver (V2.1.0 FCC firmware). The RX will bind as usual and telemetry is received and looks reasonable. The channel data are checked with the BetaFlight configurator with the RX connected to an AlienFlightNG Anubis16 flight controller. V1.x.x FCC mode is also still working as expected. The LBT (EU) mode is not yet tested. More testing and feedback is welcome. The code came partially from the MPM:

https://github.com/pascallanger/DIY-Multiprotocol-TX-Module

Edit: I really don't wanted to offend anyone here. Fur sure this work would not be possible without Pascal's efforts to reverse engineering the protocol. Many thanks go to him for the initial implementation. Since I was looking for deviation support of the V2.1.x protocol I used the opportunity and based my code on his work. I also posted in the Deviation forum to ask if anyone is already working on this before I looked into the code:

https://www.deviationtx.com/forum/protocol-development/8682-frsky-accst-v2-1-x-protocol-support

Locks to be my initial thinking to reference the source of the code was not good enough and I want to apologize for this.

### CODE REVIEW

File: src/protocol/frskyx_cc2500.c [hunk 1]
```diff
@@ -132,6 +137,36 @@ static const u16 CRCTable[] = {
 
 
 
+static void init_hop_FRSkyX2(void)
+{
+    //Increment
+    u8 inc = (fixed_id % 46) + 1;
+    if( inc == 12 || inc ==35 ) inc++;                          //Exception list from dumps
+    //Start offset
```
Comment (PhracturedBlue) [inline src/protocol/frskyx_cc2500.c:145]: Lint Error:
```
142:  Should have a space between // and comment  
144:  Should have a space between // and comment  
144:  Missing space before ( in if(  
145:  Should have a space between // and comment  
```

File: src/protocol/frskyx_cc2500.c [hunk 2]
```diff
@@ -132,6 +137,36 @@ static const u16 CRCTable[] = {
...
+    u8 offset = fixed_id % 5;
+
+    u8 channel;
+    for(u8 i=0; i<47; i++)
+    {
+        channel = 5 * ((u16)(inc * i) % 47) + offset;
+        //Exception list from dumps
+        if(Model.proto_opts[PROTO_OPTS_FORMAT]) // LBT or FCC
```
Comment (PhracturedBlue) [inline src/protocol/frskyx_cc2500.c:153]: Lint Error:
```
149:  Missing spaces around =  
149:  Missing spaces around <  
149:  Missing space before ( in for(  
152:  Should have a space between // and comment  
153:  At least two spaces is best between code and comments  
153:  Missing space before ( in if(  
```

File: src/protocol/frskyx_cc2500.c [hunk 3]
```diff
@@ -132,6 +137,36 @@ static const u16 CRCTable[] = {
...
+        {
+            //LBT
+            if( channel <=1 || channel == 43 || channel == 44 || channel == 87 || channel == 88 || channel == 129 || channel == 130 || channel == 173 || channel == 174)
+                channel += 2;
+            else if( channel == 216 || channel == 217 || channel == 218)
```
Comment (PhracturedBlue) [inline src/protocol/frskyx_cc2500.c:158]: Lint Error:
```
155:  Should have a space between // and comment  
156:  Missing space before ( in if(  
158:  Missing space before ( in if(  
```

File: src/protocol/frskyx_cc2500.c [hunk 4]
```diff
@@ -132,6 +137,36 @@ static const u16 CRCTable[] = {
...
+                channel += 3;
+        }
+        else // FCC
+            if ( channel == 3 || channel == 4 || channel == 46 || channel == 47 || channel == 90 || channel == 91  || channel == 133 || channel == 134 || channel == 176 || channel == 177 || channel == 220 || channel == 221 )
+                channel += 2;
+        //Store
```
Comment (PhracturedBlue) [inline src/protocol/frskyx_cc2500.c:164]: Lint Error:
```
161:  At least two spaces is best between code and comments  
164:  Should have a space between // and comment  
```

File: src/protocol/frskyx_cc2500.c [hunk 5]
```diff
@@ -132,6 +137,36 @@ static const u16 CRCTable[] = {
...
+        hop_data_v2[i] = channel;
+    }
+    hop_data_v2[47] = 0;                                  //Bind freq
```
Comment (PhracturedBlue) [inline src/protocol/frskyx_cc2500.c:167]: Lint Error:
```
167:  Should have a space between // and comment  
```

File: src/protocol/frskyx_cc2500.c [hunk 6]
```diff
@@ -153,33 +188,61 @@ static void set_start(u8 ch) {
   CC2500_WriteReg(CC2500_23_FSCAL3, calData[ch][0]);
   CC2500_WriteReg(CC2500_24_FSCAL2, calData[ch][1]);
   CC2500_WriteReg(CC2500_25_FSCAL1, calData[ch][2]);
-  CC2500_WriteReg(CC2500_0A_CHANNR, ch == 47 ? 0 : hop_data[ch]);
+  if(!Model.proto_opts[PROTO_OPTS_VERSION])
+      CC2500_WriteReg(CC2500_0A_CHANNR, ch == 47 ? 0 : hop_data[ch]);
+  else {
```
Comment (PhracturedBlue) [inline src/protocol/frskyx_cc2500.c:193]: Lint Error:
```
191:  Missing space before ( in if(  
193:  If an else has a brace on one side, it should have it on both  
```

File: src/protocol/frskyx_cc2500.c [hunk 7]
```diff
@@ -153,33 +188,61 @@ static void set_start(u8 ch) {
...
-    packet[11] = 0x02;
-    packet[12] = RXNUM;
-
-    memset(&packet[13], 0, packet_size-15);
-
-    u16 lcrc = crc(&packet[3], packet_size-5);
-    packet[packet_size-2] = lcrc >> 8;
-    packet[packet_size-1] = lcrc;
+
+    if(!Model.proto_opts[PROTO_OPTS_VERSION]) {
```
Comment (PhracturedBlue) [inline src/protocol/frskyx_cc2500.c:208]: Lint Error:
```
208:  Missing space before ( in if(  
```

File: src/protocol/frskyx_cc2500.c [hunk 8]
```diff
@@ -153,33 +188,61 @@ static void set_start(u8 ch) {
...
+        packet[9] = hop_data[idx++];
+        packet[10] = hop_data[idx++];
+        packet[11] = 0x02;
+        packet[12] = RXNUM;
+
+        memset(&packet[13], 0, packet_size-14);
+        if(binding_idx&0x01)
+            memcpy(&packet[13],(void *)"\x55\xAA\x5A\xA5",4);   // Telem off
+        if(binding_idx&0x02)
+            memcpy(&packet[17],(void *)"\x55\xAA\x5A\xA5",4);   // CH9-16
```
Comment (PhracturedBlue) [inline src/protocol/frskyx_cc2500.c:223]: Lint Error:
```
220:  Missing space before ( in if(  
221:  Missing space after ,  
222:  Missing space before ( in if(  
223:  Missing space after ,  
```

File: src/protocol/frskyx_cc2500.c [hunk 9]
```diff
@@ -153,33 +188,61 @@ static void set_start(u8 ch) {
...
+    } else {
+        //packet 1D 03 01 0E 1C 02 00 00 32 0B 00 00 A8 26 28 01 A1 00 00 00 3E F6 87 C7 00 00 00 00 C9 C9
+        packet[5] = 0x02;                   // Unknown but constant ID?
+        packet[6] = RXNUM;
+        //Bind flags
+        packet[7]=0;
```
Comment (PhracturedBlue) [inline src/protocol/frskyx_cc2500.c:229]: Lint Error:
```
225:  Should have a space between // and comment  
228:  Should have a space between // and comment  
229:  Missing spaces around =  
```

File: src/protocol/frskyx_cc2500.c [hunk 10]
```diff
@@ -153,33 +188,61 @@ static void set_start(u8 ch) {
...
+        if(binding_idx&0x01)
+            packet[7] |= 0x40;              // Telem off
+        if(binding_idx&0x02)
+            packet[7] |= 0x80;              // CH9-16
+        //Unknown bytes
```
Comment (PhracturedBlue) [inline src/protocol/frskyx_cc2500.c:234]: Lint Error:
```
230:  Missing space before ( in if(  
232:  Missing space before ( in if(  
234:  Should have a space between // and comment  
```

File: src/protocol/frskyx_cc2500.c [hunk 11]
```diff
@@ -153,33 +188,61 @@ static void set_start(u8 ch) {
...
+        memcpy(&packet[8],"\x32\x0B\x00\x00\xA8\x26\x28\x01\xA1\x00\x00\x00\x3E\xF6\x87\xC7",16);
+        packet[20]^= 0x0E ^ fixed_id;       // Update the ID
+        packet[21]^= 0x1C ^ fixed_id >> 8;  // Update the ID
+        //Xor
+        for(u8 i=3; i<packet_size-1; i++)
```
Comment (PhracturedBlue) [inline src/protocol/frskyx_cc2500.c:239]: Lint Error:
```
235:  Missing space after ,  
238:  Should have a space between // and comment  
239:  Missing spaces around =  
239:  Missing spaces around <  
239:  Missing space before ( in for(  
```

File: src/protocol/frskyx_cc2500.c [hunk 12]
```diff
@@ -589,13 +652,28 @@ static void frskyX_init() {
   for (u32 i=0; i < ((sizeof init_data_shared) / (sizeof init_data_shared[0])); i++)
       CC2500_WriteReg(init_data_shared[i][0], init_data_shared[i][1]);
 
+  if(Model.proto_opts[PROTO_OPTS_VERSION])
+  {
+      CC2500_WriteReg(CC2500_08_PKTCTRL0, 0x05);      // Enable CRC
+      if(!Model.proto_opts[PROTO_OPTS_FORMAT])
+      { // FCC
```
Comment (PhracturedBlue) [inline src/protocol/frskyx_cc2500.c:659]: Lint Error:
```
655:  Missing space before ( in if(  
658:  Missing space before ( in if(  
659:  At least two spaces is best between code and comments  
```

File: src/protocol/frskyx_cc2500.c [hunk 13]
```diff
@@ -589,13 +652,28 @@ static void frskyX_init() {
...
   CC2500_WriteReg(CC2500_0C_FSCTRL0, fine);
   CC2500_Strobe(CC2500_SIDLE);
 
   //calibrate hop channels
   for (u8 c = 0; c < 47; c++) {
       CC2500_Strobe(CC2500_SIDLE);
-      CC2500_WriteReg(CC2500_0A_CHANNR, hop_data[c]);
+      if(!Model.proto_opts[PROTO_OPTS_VERSION])
+          CC2500_WriteReg(CC2500_0A_CHANNR, hop_data[c]);
+      else {
```
Comment (PhracturedBlue) [inline src/protocol/frskyx_cc2500.c:673]: Lint Error:
```
671:  Missing space before ( in if(  
673:  If an else has a brace on one side, it should have it on both  
```

File: src/protocol/frskyx_cc2500.c [hunk 14]
```diff
@@ -635,7 +713,9 @@ static void initialize(int bind)
 
     // initialize statics since 7e modules don't initialize
     fine = Model.proto_opts[PROTO_OPTS_FREQFINE];
-    packet_size = Model.proto_opts[PROTO_OPTS_FORMAT] ? 33 : 30;
+    packet_size = 0x1D;
+    if(!Model.proto_opts[PROTO_OPTS_VERSION] && Model.proto_opts[PROTO_OPTS_FORMAT])
+        packet_size=0x20;               // FrSkyX V1 LBT
```
Comment (PhracturedBlue) [inline src/protocol/frskyx_cc2500.c:718]: Lint Error:
```
717:  Missing space before ( in if(  
718:  Missing spaces around =  
```

File: src/protocol/frskyx_cc2500.c [hunk 15]
```diff
@@ -132,16 +137,46 @@ static const u16 CRCTable[] = {
 
 
 
+static void init_hop_FRSkyX2(void)
+{
+    // Increment
+    u8 inc = (fixed_id % 46) + 1;
+    if ( inc == 12 || inc ==35 ) inc++;                     //Exception list from dumps
```
Comment (PhracturedBlue) [inline src/protocol/frskyx_cc2500.c:144]: Lint Error:
```
144:  Should have a space between // and comment  
```

File: src/protocol/frskyx_cc2500.c [hunk 16]
```diff
@@ -153,33 +188,60 @@ static void set_start(u8 ch) {
...
+        packet[5] = 0x02;                   // Unknown but constant ID?
+        packet[6] = RXNUM;
+        // Bind flags
+        packet[7] = 0;
+        if (binding_idx&0x01)
+            packet[7] |= 0x40;              // Telem off
+        if (binding_idx&0x02)
+            packet[7] |= 0x80;              // CH9-16
+        // Unknown bytes
+        memcpy(&packet[8],"\x32\x0B\x00\x00\xA8\x26\x28\x01\xA1\x00\x00\x00\x3E\xF6\x87\xC7",16);
```
Comment (PhracturedBlue) [inline src/protocol/frskyx_cc2500.c:234]: Lint Error:
```
234:  Missing space after ,  
```

File: src/protocol/frskyx_cc2500.c [hunk 17]
```diff
@@ -584,18 +646,34 @@ static void frskyX_init() {
...
 
-  //calibrate hop channels
+  // calibrate hop channels
   for (u8 c = 0; c < 47; c++) {
       CC2500_Strobe(CC2500_SIDLE);
-      CC2500_WriteReg(CC2500_0A_CHANNR, hop_data[c]);
+      if (!Model.proto_opts[PROTO_OPTS_VERSION]) {
+          CC2500_WriteReg(CC2500_0A_CHANNR, hop_data[c]);
+      }
+      else {
```
Comment (PhracturedBlue) [inline src/protocol/frskyx_cc2500.c:673]: Lint Error:
```
673:  If an else has a brace on one side, it should have it on both  
```

File: src/protocol/frskyx_cc2500.c [hunk 18]
```diff
@@ -635,7 +713,9 @@ static void initialize(int bind)
 
     // initialize statics since 7e modules don't initialize
     fine = Model.proto_opts[PROTO_OPTS_FREQFINE];
-    packet_size = Model.proto_opts[PROTO_OPTS_FORMAT] ? 33 : 30;
+    packet_size = 0x1D;
+    if (!Model.proto_opts[PROTO_OPTS_VERSION] && Model.proto_opts[PROTO_OPTS_FORMAT])
+        packet_size=0x20;               // FrSkyX V1 LBT
```
Comment (PhracturedBlue) [inline src/protocol/frskyx_cc2500.c:718]: Lint Error:
```
718:  Missing spaces around =  
```

File: src/protocol/frskyx_cc2500.c [hunk 19]
```diff
@@ -680,7 +760,7 @@ uintptr_t FRSKYX_Cmds(enum ProtoCmds cmd)
 {
     switch(cmd) {
         case PROTOCMD_INIT: initialize(0); return 0;
-        case PROTOCMD_CHECK_AUTOBIND: return 0; //Never Autobind
+        case PROTOCMD_CHECK_AUTOBIND: return 0; // Never Autobind
```
Comment (PhracturedBlue) [inline src/protocol/frskyx_cc2500.c:763]: Lint Error:
```
763:  At least two spaces is best between code and comments  
```

**Comment by pascallanger** [comment]
When you say "partially", all the modifications you've applied come from my code... A thank you for the long hours I've spent reversing this protocol would be appreciated...

**Comment by MJ666** [comment]
For sure the updates I applied to deviation are completely based on your code. Parts are a plain copy and other parts are rewritten and modified. I definitely appreciate the work you have done for MPM and whiteout this my work would not be possible.

**Comment by TheRealMoeder** [comment]
I'm sure MJ666 didn't try to offend you Pascal. @MJ666 Why not just add a small comment in the code referencing Pascal's MPM work in this protocol? You can find such reference in other protocol modules as well. Nonetheless: Thanks to both of you for the hard work!

**Comment by pascallanger** [comment]
That's not what I meant... I was just saying you can add a thanks note when porting a protocol to the author in the pull request. That's all...

**Comment by MJ666** [comment]
Also bench tested LBT firmware with R-XSR LBT V2.1.0 firmware. Chanel output looks to be OK but telemetry RSSI (around 120 when RX and TX near by) and LQI (static 127 not depending on distance) values looks to be strange on LBT. For the FCC firmware values lock to be OK. RSSI is around 100 and LQI around 50.

**Comment by TheRealMoeder** [comment]
I have no RX to check this. I'd prefer a some more positive feedback before merging if possible.

**Comment by MJ666** [comment]
Have tested this some more with theR-XSR all firmware variants and a updated version of Betaflight:
https://github.com/betaflight/betaflight/pull/9697#issuecomment-613673269
I successfully flight tested this with all four protocol variants. I have see some telemetry issues with BF wich is likely in the Betaflight code. Looks to be LBT v 2.1 is mostly affected. I now also have an Jumper T16 with the MPM. The telemetry values i can see with the R-XSR are very similar to what i can see with the updaded Deviation code. Values are quite different between the firmware variants for some reason. With LBT I have very high LQI values. Is If the is someone out there who requires a binary for testing please let me know your TX and I can build them.

**Comment by pascallanger** [comment]
I had to adjust the protocol timing for the telemetry (while developing before the release, nothing has changed since the release and I have no complains). Rssi/lqi is usually something but the number of packets per seconds is the true indication. For example all the telemetry looked to work but I was only receiving a third of the packets... I haven't checked your code but I also implemented LBT per the local regulation (not the protocol name but really listen before talk).
Pascal

**Comment by MJ666** [comment]
I made some comparison between DEVO10, Jumper T16 (MPM) and the four protocols. (MPM). Here are the test results:

`DEVO v1 FCC
RSSI	89-92
LQI		43-47
LRSSI	-11- -17
VOLT1	3.3V

DEVO v1 LBT
RSSI	94-96
LQI		127 statisch
LRSSI	-16- -23
VOLT1	3.3V

DEVO v2.1 FCC
RSSI	115-116
LQI		41-46
LRSSI	-21- -26
VOLT1	3.3V

DEVO v2.1 LBT (looks to be only sporadic updates)
RSSI	109-112
LQI		127 statisch
LRSSI	-16- -23
VOLT1	3.3V

T16 v1 FCC
TRSS	120-121
TQIy	42-48
RSSI	100-101
RxbT	3.5V

T16 v1 LBT
TRSS	97-100
TQIy	127 statisch
RSSI	96-99
RxbT	3.5V

T16 v2.1 FCC
TRSS	105-110
TQIy	40-46
RSSI	103-106
RxbT	3.5V

T16 v2.1 LBT
TRSS	105-110
TQIy	127 statisch
RSSI	103-106
RxbT	3.5V
`
The R-XSR is here only powered by 3.3V when connected to the FC on the bench. I also did some tests with a 5V supply but this has not affected the results. The results with the BETAFPV F4 FC 1S Brushless FC look similar. There are some difference how MPM and deviation is showing the telemetry data. I can see for V2 LBT the telemetry is very unstable with the Devo but not with the MPM. I did quite some flying with the BetaFPV last days with all different protocols including the V2 LBT and have not seen any issues at this side.

**Comment by pascallanger** [comment]
Again, you are looking at RSSI, TRSS TQly which are not good indicators. You need to look at the packets per seconds on the telemetry to have a real idea of what's going on.

**Comment by vladislavy** [comment]
arm-none-eabi-ld: objs/devo7e/frskyx_cc2500.elf section `.text' will not fit in region `ram'
arm-none-eabi-ld: region `ram' overflowed by 454 bytes
make[1]: *** [objs/devo7e/frskyx_cc2500.bin] Error 1
make[1]: *** Waiting for unfinished jobs....
make: *** [zip_devo7e] Error 2

devo7e protocol module can't fit to 4K

**Comment by MJ666** [comment]
Not sure what options we have to shrink the size of the module. Likely we need to Keep the devo7e without v2 support and add a compile time option for this?

**Comment by vladislavy** [comment]
You can use MODULAR for it.

**Comment by TheRealMoeder** [comment]
init_hop_FRSkyX2 looks pretty expensive with all its comparisons. Any ideas to shrink this?

**Comment by MJ666** [comment]
@vladislavy updated the PR according to your advice. 

@TheRealMoeder this code allows the bind information to be build in the proven way. Since this algorytm results in a table depending on the variable ID on TX and RX side. Not sure how this can be simplified without breaking something. A separate module would be possible but will impact the usability for all Deviation users and is duplication a lot of code. The image size for all transmitter will grow even more.

**Comment by TheRealMoeder** [comment]
There should be more room for optimization on the code hop_data array can also be shared for example. I'll try to have a look into it

**Comment by TheRealMoeder** [comment]
Looking at @pascallanger 's MPM code, he generates most part of the CRC table. Reusing the same code we can save 464 bytes in my quick test. Didn't check functionality, just copied the function, but I don't see why it shouldn't.

**Comment by TheRealMoeder** [comment]
Another 84 bytes saved with:

        // Exception list from dumps
        if (Model.proto_opts[PROTO_OPTS_FORMAT])                // LBT or FCC
        {
            // LBT
            for (u8 k = 0; k < 1; k++) {
                if (channel <= 1 || channel == 43+k || channel == 87+k || channel == 129+k || channel == 173 +k)
                    channel += 2;
                else if (channel == 216+k || channel == 218)
                    channel += 3;
            }
        }
        else
            // FCC
            for (u8 k = 0; k < 1; k++) {
                if (channel == 3+k || channel == 46+k || channel == 90+k || channel == 133+k || channel == 176+k || channel == 220+k)
                    channel += 2;
            }
        hop_data_v2[i] = channel;                               // Store

File: src/protocol/frskyx_cc2500.c [hunk 20]
```diff
@@ -95,53 +102,60 @@ static const u8 hop_data[] = {
...
+  val /= 16;
+  return word ^ (0x1081 * val);
+}
+
 static u16 crc(u8 *data, u8 len) {
   u16 crc = 0;
-  for(int i=0; i < len; i++)
-      crc = (crc<<8) ^ CRCTable[((u8)(crc>>8) ^ *data++) & 0xFF];
+  for (int i = 0; i < len; i++)
+      crc = (crc<<8) ^ crcTable((u8)(crc>>8) ^ *data++);
```
Comment (PhracturedBlue) [inline src/protocol/frskyx_cc2500.c:152]: Lint Error:
```
152:  Missing spaces around <<  
```

**Comment by MJ666** [comment]
Replaced CRC calculation code and it fits to the Devo7e module now. Thanks to @TheRealMoeder for the heads up and @pascallanger forr the smaller code size.

File: src/protocol/frskyx_cc2500.c [hunk 21]
```diff
@@ -584,18 +625,33 @@ static void frskyX_init() {
...
+          CC2500_WriteReg(CC2500_11_MDMCFG3, 0x84);   // bitrate 70K->77K
+      }
+  }
+
   CC2500_WriteReg(CC2500_0C_FSCTRL0, fine);
   CC2500_Strobe(CC2500_SIDLE);
 
-  //calibrate hop channels
+  // calibrate hop channels
   for (u8 c = 0; c < 47; c++) {
```
Comment (TheRealMoeder) [inline src/protocol/frskyx_cc2500.c:647]: Please replace magic number (HOP_DATA_SIZE - 1)

**Comment by pascallanger** [comment]
@goebish The Multi FrSKy_RX can already auto bind D8, D16 v1 (FCC/LBT) and D16v2.1 (FCC/LBT). I've also added a CloneTX mode which enables someone to duplicate the ID&Hop of a TX sending FrSky for all 3 modes.
Pascal

**Comment by MJ666** [comment]
I rebased now and tested the updated CRC code with a single protocol. All protocol variants are fligth tested before this change a few times with the BetaFPV FC. I also have bench tested all protocols with a original FrSky R-XSR receiver connect to an AlienFlightNG FC and the BF configurator.

File: src/protocol/frskyx_cc2500.c (hunk 21 above)
Comment (MJ666) [reply src/protocol/frskyx_cc2500.c]: Done.

**Comment by vladislavy** [comment]
You need correct HOP_DATA_SIZE or HOPP_DATA_SIZE.

**Comment by MJ666** [comment]
> You need correct HOP_DATA_SIZE or HOPP_DATA_SIZE.

Sorry my bad. Fixed.

**Comment by vladislavy** [comment]
Now devof7 protocol module doesn't fit, overflowed by 6 bytes.

**Comment by MJ666** [comment]
Mhhh, to big again for Devo7. But still fits here when I compile localy.

**Comment by TheRealMoeder** [comment]
This loop saves enough space. Please double check channel calculation!
static void init_hop_FRSkyX2(void)
{
    u8 inc = (fixed_id % 46) + 1;                               // Increment
    if ( inc == 12 || inc == 35 ) inc++;                        // Exception list from dumps
    // Start offset
    u8 offset = fixed_id % 5;

    u8 channel;
    for (u8 i = 0; i < 47; i++)
    {
        channel = 5 * ((u16)(inc * i) % 47) + offset;
        // Exception list from dumps
        for (u8 k = 0; k < 2; k++) {
            if (Model.proto_opts[PROTO_OPTS_FORMAT])                // LBT or FCC
            {
                // LBT
                if (channel <= 1 || channel == 43+k || channel == 87+k || channel == 129+k || channel == 173 +k)
                    channel += 2;
                else if (channel == 216+k || channel == 218)
                    channel += 3;
            }
            else
                // FCC
                if (channel == 3+k || channel == 46+k || channel == 90+k || channel == 133+k || channel == 176+k || channel == 220+k)
                    channel += 2;
        }
        hop_data_v2[i] = channel;                               // Store
    }
    hop_data_v2[47] = 0;                                        // Bind freq
}

**Comment by MJ666** [comment]
This is what im actually working on. dont realy understand why the complex code need less space but anyhow.... It fits but I need to retest the V2 protocols. :-) Stay tuned.

**Comment by MJ666** [comment]
After cleanup and some more fixing it finaly fits all targets with full functionality. I have done again some testing with R-XSR, XM+ and the BetaFPV FC. With the BetaFPV whoop i did sucessful flight tests with all four protocols. There is still a bind problem with LBTv1 and the BF test code (https://github.com/betaflight/betaflight/pull/9697) im using but this is also there with the Jumper t16 (MPM). Dont have an orginal FrSky TX fro testing. Since hopping tables are identical with FCCv1 you can fly without issues. Binding is working for all Protocols with original FrSky receivers. Bench tests with them dont show any problems. Unfortunately all my FrSky receivers are soldered into my small quadcopters and are not easily can be updated.

File: src/protocol/frskyx_cc2500.c [hunk 22]
```diff
@@ -95,53 +102,58 @@ static const u8 hop_data[] = {
...
+            else if (channel == 216 || channel == 217 || channel == 218)
+                channel += 3;
+        } else {
+            // FCC
+            if (channel == 3 || channel == 4 || channel == 46 || channel == 47 || channel == 90 || channel == 91  || channel == 133 || channel == 134 || channel == 176 || channel == 177 || channel == 220 || channel == 221)
+                channel += 2;
+        }
+        hop_data_v2[i] = channel;                               // Store
+    }
+    hop_data_v2[47] = 0;                                        // Bind freq
```
Comment (TheRealMoeder) [inline src/protocol/frskyx_cc2500.c:136]: This could be HOP_DATA_SIZE as well.

Comment (MJ666) [reply src/protocol/frskyx_cc2500.c]: Done.

**Comment by TheRealMoeder** [comment]
@MJ666 Could you also update deviation-manual with the new protocol options?

**Pull request merged**